File: demo/services/flaky.py
"""A service that stays up until you ask it not to.

`./demo crash` drops a flag file and this exits 42, so the sidebar shows a
failed service with a real exit code. `./demo revive` brings it back under a
new container ID, which is the reattach path.
"""

import os
import random
import sys
import time

sys.path.insert(0, os.path.dirname(__file__))
from common import (  # noqa: E402
    GREEN, GREY, RED, RESET, YELLOW, flag, install_clean_exit, jitter, log,
)

EXIT_CODE = 42


def main():
    install_clean_exit()
    log("INFO", "flaky-sync starting")
    log("INFO", f"{GREEN}ready{RESET} - syncing every ~3s")
    n = 0

    while True:
        if flag("crash"):
            log("ERROR", f"{RED}fatal: replication slot lost, cannot continue{RESET}")
            log("ERROR", f"{GREY}  last good offset: {random.randint(10**7, 10**8)}{RESET}")
            log("FATAL", f"{RED}exiting with code {EXIT_CODE}{RESET}")
            sys.stdout.flush()
            os._exit(EXIT_CODE)

        n += 1
        if n % 5 == 0:
            log("WARN", f"{YELLOW}sync lagging: {random.randint(200, 4000)}ms behind primary{RESET}")
        else:
            log("INFO", f"synced {random.randint(1, 90)} rows {GREY}(batch {n}){RESET}")
        time.sleep(jitter(3.0))


if __name__ == "__main__":
    main()

File: demo/services/flaky.ts
/**
 * A service that stays up until you ask it not to.
 *
 * `./stack crash` drops a flag file and this exits 42, so the sidebar shows a
 * failed service with a real exit code. `./stack revive` brings it back under a
 * new container ID, which is the reattach path.
 */

import {
  GREEN, GREY, RED, RESET, YELLOW,
  exitAfterFlush, flag, installCleanExit, jitter, log, randint, sleep,
} from "./common.ts";

const EXIT_CODE = 42;

async function main(): Promise<void> {
  installCleanExit();
  log("INFO", "flaky-sync starting");
  log("INFO", `${GREEN}ready${RESET} - syncing every ~3s`);
  let n = 0;

  for (;;) {
    if (flag("crash")) {
      log("ERROR", `${RED}fatal: replication slot lost, cannot continue${RESET}`);
      log("ERROR", `${GREY}  last good offset: ${randint(10 ** 7, 10 ** 8)}${RESET}`);
      log("FATAL", `${RED}exiting with code ${EXIT_CODE}${RESET}`);
      exitAfterFlush(EXIT_CODE);
      return;
    }

    n += 1;
    if (n % 5 === 0) {
      log("WARN", `${YELLOW}sync lagging: ${randint(200, 4000)}ms behind primary${RESET}`);
    } else {
      log("INFO", `synced ${randint(1, 90)} rows ${GREY}(batch ${n})${RESET}`);
    }
    await sleep(jitter(3.0));
  }
}

await main();
